pkg/util/manager/manager.go

```diff
@@ -22,40 +22,83 @@ import (
 	"github.com/kubesphere/kubekey/pkg/util/runner"
 	log "github.com/sirupsen/logrus"
+	"sync"
+)
+
+const (
+	IsInitCluster = true
+	Docker        = "docker"
+	Conatinerd    = "containerd"
+	Crio          = "crio"
+	Isula         = "isula"
 )
 
 // Manager defines all the parameters needed for the installation.
 type Manager struct {
-	ObjName            string
-	Cluster            *kubekeyapiv1alpha1.ClusterSpec
-	Logger             log.FieldLogger
-	Connector          connector.Connector
-	Runner             *runner.Runner
-	AllNodes           []kubekeyapiv1alpha1.HostCfg
-	EtcdNodes          []kubekeyapiv1alpha1.HostCfg
-	MasterNodes        []kubekeyapiv1alpha1.HostCfg
-	WorkerNodes        []kubekeyapiv1alpha1.HostCfg
-	K8sNodes           []kubekeyapiv1alpha1.HostCfg
-	EtcdContainer      bool
-	ClusterHosts       []string
-	WorkDir            string
-	KsEnable           bool
-	KsVersion          string
-	Debug              bool
-	SkipCheck          bool
-	SkipPullImages     bool
-	SourcesDir         string
-	AddImagesRepo      bool
-	InCluster          bool
-	DeployLocalStorage bool
-	Kubeconfig         string
-	Conditions         []kubekeyapiv1alpha1.Condition
-	ClientSet          *kubekeyclientset.Clientset
-	DownloadCommand    func(path, url string) string
+	ObjName                  string
+	Cluster                  *kubekeyapiv1alpha1.ClusterSpec
+	Logger                   log.FieldLogger
+	Connector                connector.Connector
+	Runner                   *runner.Runner
+	AllNodes                 []kubekeyapiv1alpha1.HostCfg
+	EtcdNodes                []kubekeyapiv1alpha1.HostCfg
+	MasterNodes              []kubekeyapiv1alpha1.HostCfg
+	WorkerNodes              []kubekeyapiv1alpha1.HostCfg
+	K8sNodes                 []kubekeyapiv1alpha1.HostCfg
+	EtcdContainer            bool
+	ClusterHosts             []string
+	WorkDir                  string
+	KsEnable                 bool
+	KsVersion                string
+	Debug                    bool
+	SkipCheck                bool
+	SkipPullImages           bool
+	SourcesDir               string
+	AddImagesRepo            bool
+	InCluster                bool
+	ContainerManager         string
+	ContainerRuntimeEndpoint string
+	DeployLocalStorage       bool
+	Kubeconfig               string
+	ClusterStatus            *ClusterStatus
+	UpgradeStatus            *UpgradeStatus
+	Conditions               []kubekeyapiv1alpha1.Condition
+	ClientSet                *kubekeyclientset.Clientset
+	DownloadCommand          func(path, url string) string
+}
+
+// ClusterStatus is used to store cluster status
+type ClusterStatus struct {
+	IsExist        bool
+	Version        string
+	AllNodesInfo   map[string]string
+	Kubeconfig     string
+	BootstrapToken string
+	CertificateKey string
+}
+
+// UpgradeStatus is used to store cluster upgrade status
+type UpgradeStatus struct {
+	CurrentVersions   map[string]string
+	CurrentVersionStr string
+	NextVersionStr    string
+	MU                sync.Mutex
```

Comment:
The mgr will deep copy itself when exec task parallel, so the mutex should be define as a pointer. Otherwise, i think it will generate a new mutex after mgr.Copy().

Reply:
Done